### ISSUE TITLE
After upgrading to the latest version it picks the Windows backend now and then (not always)

### ISSUE THREAD

**Issue by jaraco** [jaraco/keyring#112]
And the system is OS X.

```
#!bash

  File "/usr/local/lib/python2.7/site-packages/mercurial_keyring.py", line 380, in find_user_password
    return self._pwd_handler.find_auth(self, realm, authuri, req)
  File "/usr/local/lib/python2.7/site-packages/mercurial_keyring.py", line 231, in find_auth
    pwd = password_store.get_http_password(keyring_url, user)
  File "/usr/local/lib/python2.7/site-packages/mercurial_keyring.py", line 114, in get_http_password
    self._format_http_key(url, username))
  File "/usr/local/lib/python2.7/site-packages/keyring/core.py", line 36, in get_password
    return _keyring_backend.get_password(service_name, username)
  File "/usr/local/lib/python2.7/site-packages/keyring/backends/Windows.py", line 96, in get_password
    res = self._get_password(service)
  File "/usr/local/lib/python2.7/site-packages/keyring/backends/Windows.py", line 111, in _get_password
    except pywintypes.error:
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 86, in __getattribute__
    self._load()
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 58, in _load
    mod = _origimport(head, globals, locals)
ImportError: No module named pywintypes
abort: No module named pywintypes!
```

---
- Bitbucket: https://bitbucket.org/kang/python-keyring-lib/issue/112
- Originally reported by: Jeffrey Gelens
- Originally created at: 2013-08-27T09:50:18.572

**Comment by jaraco** [comment]
Lame! I believe that's due to Mercurial's Demand Import feature (which makes it hard to detect presence or absence of modules). I'll work out a solution soon. In the meantime, feel free to use the config files to specify a back end or use the last 1.x release.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Backend viability/priority checks now are more aggressive about module presence checking, requesting **name** from imported modules to force the demand importer to actually attempt the import. Fixes #112.

→ <<cset 55831944f1c0>>

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Thanks for the report and sorry about the troubles. Please try 2.0.3 and let me know if that doesn't resolve the issue.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
It's working great.
Thanks for the quick patch! 

---

Original comment by: Jeffrey Gelens

**Comment by jaraco** [comment]
While 2.0.3 was working great, versions released after that create the following error with mercurial_keyring:

```
#!bash
*** failed to import extension mercurial_keyring: 'module' object has no attribute 'system'
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/mercurial/extensions.py", line 92, in loadall
    load(ui, name, path)
  File "/usr/local/lib/python2.7/site-packages/mercurial/extensions.py", line 79, in load
    mod = importh(name)
  File "/usr/local/lib/python2.7/site-packages/mercurial/extensions.py", line 69, in importh
    mod = __import__(name)
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 95, in _demandimport
    return _import(name, globals, locals, fromlist, level)
  File "/usr/local/lib/python2.7/site-packages/mercurial_keyring.py", line 82, in <module>
    import keyring
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 95, in _demandimport
    return _import(name, globals, locals, fromlist, level)
  File "/usr/local/lib/python2.7/site-packages/keyring/__init__.py", line 12, in <module>
    from .core import (set_keyring, get_keyring, set_password, get_password,
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 112, in _demandimport
    return _origimport(name, globals, locals, fromlist, level)
  File "/usr/local/lib/python2.7/site-packages/keyring/core.py", line 166, in <module>
    init_backend()
  File "/usr/local/lib/python2.7/site-packages/keyring/core.py", line 58, in init_backend
    keyring = load_config()
  File "/usr/local/lib/python2.7/site-packages/keyring/core.py", line 127, in load_config
    config_path = os.path.join(platform.data_root(), filename)
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 86, in __getattribute__
    self._load()
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 58, in _load
    mod = _origimport(head, globals, locals)
  File "/usr/local/lib/python2.7/site-packages/keyring/util/platform.py", line 24, in <module>
    data_root = globals().get('_data_root_' + platform.system(), _data_root_Linux)
  File "/usr/local/lib/python2.7/site-packages/mercurial/demandimport.py", line 87, in __getattribute__
    return getattr(self._module, attr)
AttributeError: 'module' object has no attribute 'system'

[255] » pip freeze | grep keyring                                                                           ~/C/S/P/src 255
keyring==3.0.1
mercurial-keyring==0.6.0


```

---

Original comment by: Jeffrey Gelens

**Comment by jaraco** [comment]
Also [reported upstream](/Mekk/mercurial_keyring/issue/31). I'd like to see if Mercurial Demand Import can be altered to honor absolute_import.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
I've also filed [another report further upstream](http://bz.selenic.com/show_bug.cgi?id=4029).

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Renamed util.platform to util.platform_. Fixes #112 (part two).

→ <<cset b2629a6ef04f>>

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
The patch to fix root cause of this issue is imported into main
repository of Mercurial itself.

http://selenic.com/repo/hg/rev/e3a5922e18c3

2.8-rc or later releases should include this fix.

---

Original comment by: Katsunori FUJIWARA
